x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java

```diff
@@ -259,6 +281,119 @@ public Ccr(final Settings settings) {
         return Collections.singletonList(new FixedExecutorBuilder(settings, CCR_THREAD_POOL_NAME, 32, 100, "xpack.ccr.ccr_thread_pool"));
     }
 
+    @Override
+    public Map<String, Repository.Factory> getRepositories(Environment env, NamedXContentRegistry namedXContentRegistry) {
+        Repository.Factory repositoryFactory = (metadata) -> new CcrRepository(metadata, settings);
+        return Collections.singletonMap(CcrRepository.TYPE, repositoryFactory);
+    }
+
     protected XPackLicenseState getLicenseState() { return XPackPlugin.getSharedLicenseState(); }
 
+    private static class CCRRepositoryManager extends RemoteClusterAware implements LocalNodeMasterListener {
+
+        private static final Logger LOGGER = LogManager.getLogger(CCRRepositoryManager.class);
+        private static final String SOURCE = "refreshing " + CcrRepository.TYPE + " repositories";
+
+        private final ClusterService clusterService;
+        private final Set<String> clusters = ConcurrentCollections.newConcurrentSet();
+        private volatile boolean isMasterNode = false;
+
+        private CCRRepositoryManager(Settings settings, ClusterService clusterService) {
+            super(settings);
+            this.clusterService = clusterService;
+            clusters.addAll(buildRemoteClustersDynamicConfig(settings).keySet());
+            clusterService.addLocalNodeMasterListener(this);
+            listenForUpdates(clusterService.getClusterSettings());
+        }
+
+        @Override
+        protected Set<String> getRemoteClusterNames() {
+            return clusters;
+        }
+
+        @Override
+        protected void updateRemoteCluster(String clusterAlias, List<String> addresses, String proxyAddress) {
+            if (addresses.isEmpty()) {
+                if (clusters.remove(clusterAlias) && isMasterNode) {
+                    refreshCCRRepositories();
+                }
+            } else {
+                if (clusters.add(clusterAlias) && isMasterNode) {
+                    refreshCCRRepositories();
+                }
+            }
+        }
+
+        @Override
+        public void onMaster() {
+            this.isMasterNode = true;
```

Comment:
instead of caching this - why not ask the cluster service if the current node is master every time we want to know? we already have a reference to it.